[자동차 경주 리팩터링] 악어 미션 제출합니다.

Hogu59 opened the pull request:
제이든님 안녕하세요,
step1에서 리뷰해주신 내용 수정해보았습니다!

말씀해주신 대로, Car 객체 생성시 이름값을 사전에 검증함으로써 Car 와 Validation 간의 의존성 제거를 해보았습니다.
많이 가르쳐주셔서 감사합니다.
혹시 추가적으로 수정하면 좋을 부분을 알려주시면 감사하겠습니다! 😃



이전 PR 링크
https://github.com/woowacourse/kotlin-racingcar/pull/108
---
`nameList` 와 `name`의 유효성 검사를 분리한 내용을 문의해주신 것에 대해서 이전 PR에도 적어두었지만 편의를 위해 다시 한번 적습니다.
```
유효성 검증을 두 부분으로 나눠서 하면 함수를 각 기능별로 분리할 수 있을것 같아서 그렇게 했었습니다!
두 부분에 있어서 발생 가능한 예외상황이 다르다고 생각해서 전체 입력인 이름 목록과 각 이름에 대한 유효성 검증을 나누었습니다!

nameList의 경우 입력된 이름 목록 자체가 유효한지 검증하려고 했습니다.

예상 오류
입력 자체가 공백 ("")
이름이 하나만 있는 경우(차가 한대이면 레이싱 경주가 불가
자동차 대수 확인을 통해 이름 목록 크기가 2 미만 20초과에 대해서는 예외처리
중복값 확인 등
name의 경우에 대해서는 각 이름이 유효성을 검증하려고 했습니다.

예상 오류
이름이 공백 (ex. "pobi", "", "crong")
이름이 5자 초과 (ex. "crongcrong")
이름에 공백이 있는 경우 등(ex. "po bi")
```
---

Hogu59 committed:
refactor(validation): Seperate car name validation

Remove dependency between Validation and Car classes.
Init block of Car class deleted.
Instead, same part was executed before making instances.
The name argument for car class is already validated, so it no longer needs any validation more in class.

Hogu59 committed:
fix(test): car name test fixed

Error occured due to seperation of the validation parts of name.
Validation no longer executed in constructor, Validation object were used instead of Car object.

Hogu59 committed:
style: execute ktlint

ktlintFormat executed.

Hogu59 committed:
docs: check the modified list

Lists were checked which are modified.

KwonDae commented on src/main/kotlin/racingcar/model/Car.kt:
```diff
@@ -1,13 +1,8 @@
-
     private fun getRandomNumber(): Int = Random.nextInt(MIN_RANDOM_NUMBER, MAX_RANDOM_NUMBER)
 
     fun move() {
```
<img width="664" alt="image" src="https://github.com/woowacourse/kotlin-racingcar/assets/33443660/df84a5b8-d7d9-4ecf-a011-c3d26f240bc9">
<img width="1075" alt="image" src="https://github.com/woowacourse/kotlin-racingcar/assets/33443660/93cfd200-5278-435b-9adf-53a2282fd1fe">


이 프로젝트에서 가장 핵심적인 비즈니스 로직인 움직이는 행위인
`move()` 메소드에 대한 테스트 코드가 없네요
이외에도 domain(model)의 로직들에 대해서 테스트 가능한 부분들은 테스트 코드 작성해주세요!

Hogu59 replied:
`move()` 메소드에 대한 테스트 코드를 작성해보았습니다.
테스트 자체가 domain(model)의 로직에 대해서 구현된다는 점을 제대로 몰랐고, 말씀해주신 덕분에 더 알게 되었습니다.
다만, 모델에 맞는 테스트코드를 작성해야한다는 전제에 대해서 고민해보면서 model을 조금 더 늘리게 되었고, controller 에 대한 테스트를 삭제하고 각 모델에 맞는 테스트를 따로 분리하여 작성하게 되었습니다.

그러다보니, 유효값 검증 자체가 init 블럭 안에서 진행하는 것이 맞다고 판단되어 옮기게 되었는데, 이렇게 구현하는 것이 바람직한지에 대한 판단인지에 대한 궁금증이 있는데, 리뷰어님께서는 어떻게 생각하시는 지 궁금합니다!

KwonDae replied:
두가지 방법이 있을것 같습니다.
1. 모델을 생성하기 전 값 검증
2. 모델을 초기화하는 Init 에서 검증

사실 정답은 없다고 생각합니다. 다만, 저의 개인적인 성향은 모델은 비즈니스 로직에 집중해야하기에 검증에 관련된 로직은 모델에서는 하지 않고, 검증된 값이 들어가야한다고 생각합니다.
다만, 어떤 모델이냐에 따라 다를 수 있기에 이 부분은 악어도 미션을 진행하면서 꾸준히 고민해보았으면 좋겠습니다.
좋은 고민입니다! 👍

KwonDae commented on src/main/kotlin/racingcar/controller/Race.kt:
Controller역할을 하고있는 Race에 대해서 한번 고민해볼까요?

<img width="1075" alt="image" src="https://github.com/woowacourse/kotlin-racingcar/assets/33443660/4e231879-4628-400e-abf4-34a2878a03df">

Controller 는 Domain(model)과 상호작용을 합니다.
그리고 domain에는 여러개의 경주차량을 관리하는 경주장(RacingGame)과 여러개의 차량(car)들로 이루어져 있죠.

그런데 지금은 경주장이 아닌 컨트롤타워에서 경주에 대한 상태, 우승자 판단 같은 것들을 혼자 다 맡아서 하고 있습니다.
때문에 테스트에서도 domain 비즈니스 로직 검증이 아닌 race 의 함수를 호출하고 있죠

즉, 명령은 Controller 담당입니다.
경주장에 대한 상태 관리와 들어온 명령에 대한 판단은 경주장 몫입니다.
차량은 자신이 얼마나 움직일지, 얼만큼 움직였는지에 대한 책임을 갖습니다.

한번 Controller 내부 책임을 분리해볼까요?

Hogu59 replied:
위의 댓글과 같이, controller의 책임 분리를 위해 model 을 분리하고, 각 모델에서 필요한 작업을 수행하도록 분리해보았습니다!
필요한 모델이 어떤것이 있는지, 더 세분화될 수는 있는지 더 많은 고민을 해야한다는 것을 알게 된 것 같습니다.
감사합니다!

KwonDae reviewed (changes requested):
안녕하세요 악어~!

2단계 미션도 잘 해주시느라 고생하셨습니다 👍 
생각해볼만한 코멘트들 남겨 드렸는데 꼭 확인하시고
수정후에 다시 요청 부탁드립니다!
끝까지 화이팅입니다 ㅎㅎ

KwonDae commented:
> Q)
> 큰 프로젝트에서는 다양한 예외상황들이 존재할 것이라고 생각하는데, 그런 경우에 있어서는 유효성 검증을 중복으로도 하는 경우가 있는지 궁금합니다! (ex. InputView에서 입력값을 1차 검증 -> Controller 에서 2차 검증)

정말 제 개인적인 경험으로는 중복 검증을 하는 경우는 거의 없었던거 같습니다.. 단지 하나의 검증으로도 확실하게끔 보장하는게 좋다고 생각합니다. 때문에 해당 검증에 대한 테스트 코드도 많이 작성했었습니다.(ex. 날짜 계산, d-day, 요일 계산 등등)

> Q)
> 실제로 유효성 검증이 중복으로도 사용된다면 보통 어디에서 이루어지는지 궁금합니다! (ex, view-controller, controller-model, view-controller-model 등등.)

중복으로도 사용된다면.. 
일단 Input 이 일어난 장소에서 한번 검증을 합니다(현재로선 InputView에서 입력값을 받은 후, split 하고나서 return 하기 전 flow 에서 첫번째 검증)
그리고 Car 객체를 만들 때에도 (ex. init 함수에서 검증) 해당 명세에 따른 검증을 실행할 수 있을것 같네요.

KwonDae commented on src/main/kotlin/racingcar/model/Car.kt:
```diff
@@ -1,13 +1,8 @@
-        Validation().carName(name)
-    }
-
     private fun getRandomNumber(): Int = Random.nextInt(MIN_RANDOM_NUMBER, MAX_RANDOM_NUMBER)
```
https://techcourse.woowahan.com/s/Q0BXzOWp/ls/hWTE4dpQ
위 구조에서 보면 랜덤값을 가져오는 로직이 Domain(model)에 종속되어야 할까요?

지금 랜덤으로 숫자를 가져오는 부분이 kotlin.random.Random 에 의존하고 있어 해당 `getRandomNumber` 함수와 이를 사용하는 곳에서는 테스트가 어렵습니다.

만약 정해진 숫자가 나온다면 함수 호출 후의 상태 변화 혹은 결과가 예측되기에 테스트를 쉽게 작성할 수 있는데요,
어떻게 하면 Car 모델에서 `getRandomNumber` 메소드가 Random 에 의존하지 않고 테스트 가능하게끔 할 수 있을까요?



이럴떄 추상화라는 개념을 사용하곤 합니다. 실제로 domain 에서는 추상화를 통한 로직을 호출하고, 이를 통해 의존성을 분리하여 테스트를 가능하게끔 할 수 있죠. (interface를 테스트하는곳에서 override를 할 수 있다!)

우리는 여기서 [Functional Interface](https://kotlinlang.org/docs/fun-interfaces.html)를 활용해볼 수 있는데요
바로 아래 피드백과 연관성이 크네요. 함꼐 고민해보시고 어려운점 있으시다면 언제든 편하게 질문주세요!

추가로, random 요소들을 테스트 하기 쉬운 구조로 개선하는 방법에 대해
아래 패턴에 대해서 참고해보시면 좋을것 같습니다 👍
[전략패턴1](https://victorydntmd.tistory.com/292)
[전략패턴2](https://refactoring.guru/ko/design-patterns/strategy)

Hogu59 replied:
`getRandomNumber`자체가 모델과 의존성이 없다는 것을 깨닫고 다른 패키지로 따로 분리하였습니다.
사용하는 로직이 과연 모델과 의존성이 존재하는가에 대해 좀 더 고민해보고 공부해봐야할 것 같습니다!

알려주신 내용 다시 참고하여 더 공부해보도록 하겠습니다!
감사합니다!

Hogu59 committed:
refactor: make move function possible to test

Make move function possible to test.
parameter was added.

Hogu59 committed:
refactor: Seperate model from controller

Seperation of models which is possible.
If tests are needed, then it should be "Model", not "Controller".

KwonDae commented on src/main/kotlin/racingcar/model/Cars.kt:
```diff
@@ -0,0 +1,9 @@
+
+import racingcar.validator.Validation
+
+class Cars(names: List<String>) {
+    init {
+        Validation().carNameList(names)
+    }
```
의도를 잘 파악해주시고 분리를 해주셨네요!
지금 Cars, RoundNumber, Winner 같은 클래스들로 분리가 되었는데 이들의 공통점은 `자동차 경주` 라는 토픽에 관심사가 있다는 것입니다.

이들이 ex. RacingGame 이라는 도메인 모델안에 들어가 관리가 된다면 어떨까요?
RacingGame 이라는 경주장 주체가 자동차, 승패여부 같은 경주에 관련된 상태를 관리하는 책임이 있기에 이들을 묶을 수 있을 것 같네요!

KwonDae commented on src/test/kotlin/racingcar/RoundNumberTest.kt:
```diff
@@ -0,0 +1,57 @@
+        Assertions.assertThat(exception.message.equals(ERROR_INPUT_NUMBER_ONLY))
+    }
+
+    @ParameterizedTest
```
ParameterizedTest 활용 👍 👍

KwonDae reviewed (approved):
이전에 코멘트 드린 부분들을 잘 이해해주시고 반영해주셨네요 👍 
분리한 클래스들에 대해서 코멘트를 남겨드렸으니 이부분은 확인해주시고
머릿속으로라도 한번 그려봤으면 해요

첫 미션 함께하시느라 고생하셨고 진행하면서 저도 즐거웠습니다 🙂
추가로 더 궁금하신게 있으시다면 언제든 편하게 DM 주세요!
고생많으셨습니다!
